src/Rbac.php

```diff
@@ -82,29 +84,109 @@ public function hasRole($role) : bool
+            if (! is_array($children)) {
+                return $rv ? true : false;
+            } elseif (! count($children)) {
+                if (is_object($needle) && (get_class($needle) == 'Zend\Permissions\Rbac\Role')) {
+                    if ($obj->getName() == $needle->getName()) {
+                        $rv++;
+                    }
+                }
+            } else {
+                foreach ($children as $child) {
+                    $roleName = $child->getName();
+                    if (is_string($needle)) {
+                        if ($roleName === $needle) {
+                            $rv++;
+                        }
+                    } elseif (is_object($needle) && (get_class($needle) == 'Zend\Permissions\Rbac\Role')) {
+                        if ($roleName == $needle->getName()) {
+                            $rv++;
+                        }
+                    }
+                    $rv += $this->roleSearchIncludingChildren($child, $needle);
+                }
+            }
         }
 
-        $roleName = $role->getName();
-        return isset($this->roles[$roleName])
-            && $this->roles[$roleName] === $role;
+        return $rv ? true : false;
     }
 
     /**
      * Get a registered role by name
      *
      * @throws Exception\InvalidArgumentException if role is not found.
      */
-    public function getRole(string $roleName) : RoleInterface
+    public function getRole(string $needle) : RoleInterface
     {
-        if (! isset($this->roles[$roleName])) {
-            throw new Exception\InvalidArgumentException(sprintf(
-                'No role with name "%s" could be found',
-                $roleName
-            ));
+        // tricky thing here is that $this->roles are an array of RoleInterface objects
+        foreach ($this->roles as $role) {
+            if ($role->getName() == $needle) {
+                return $role;
+            } else {
+                $role = $this->getRoleSearchingChildren($role, $needle);
+                if ($role != null) {
+                    return $role;
+                }
+            }
+        }
+
+        throw new Exception\InvalidArgumentException(sprintf(
+            'No role with name "%s" could be found',
+            $needle
+        ));
+    }
+
+    /**
+     * @param $obj RoleInterface
+     * @param $needle String
+     * @return null|RoleInterface
+     */
+    private function getRoleSearchingChildren($obj, $needle)
+    {
+        if (($obj instanceof RoleInterface) && ($obj->getName() == $needle)) {
+            return $obj;
+        } else {
+            $children = $obj->getChildren();
+            if (is_array($children) && ($children != null)) {
+                $result = '';
+                foreach ($children as $child) {
+                    $result = $this->getRoleSearchingChildren($child, $needle);
+                }
+                return $result;
```

Comment:
This part is confusing - do you want to return the first match, or the last? If the last, why?